Trivia: New game, general improvements

CheeseMuffin opened the pull request:
Overall changes (these are all approved by Azuuli, RO of trivia):
- Add a new game (Weakest Link), with its own question database and corresponding commands
- `/trivia qs` now scrolls so it doesn't flood the screen
- Roommods in qw can see the answers to questions
- `/trivia ladder` now scrolls with 100 people at a time, and lets people see ranks besides the top 100
- Use template strings everywhere
- Replace `/help trivia` with `/trivia help [category]`, so that people can see help on more specific things instead of flooding the screen

Tagging @Morfent

Morfent commented:
>Add a new game (Weakest Link), with its own question database and corresponding commands
I really don't like the way this is implemented. I'll make specific comments on why when I can look over it more closely

>Roommods in qw can see the answers to questions
I need to talk to Azuuli about a compromise to how this should be handled. There are ways to allow mods to more easily edit questions that require being able to see the answers to them while making it more difficult to abuse

>Replace /help trivia with /trivia help [category], so that people can see help on more specific things instead of flooding the screen
`/trivia help` would be a good alias to have, but removing `/help trivia` doesn't follow the setup help commands have in the rest of the sim. Something like how `/help tours` is handled would be a better approach

The rest of the changes look good.

CheeseMuffin commented:
> /trivia help would be a good alias to have, but removing /help trivia doesn't follow the setup help commands have in the rest of the sim. Something like how /help tours is handled would be a better approach

Basic commands in /trivia help, with a detailed smogon post?

panpawn commented on chat-plugins/trivia.js:
```diff
@@ -450,8 +451,8 @@ class Trivia extends Rooms.RoomGame {
 
-		buffer += Chat.escapeHTML(winner.name) + ' won the game with a final score of <strong>' + winner.points + '</strong>, ' +
-			'and their leaderboard score has increased by <strong>' + this.prize + '</strong> points!';
+		buffer += `${Chat.escapeHTML(winner.name)} won the game with a final score of <strong>${winner.points}</strong>, ` +
```
This should use ``Chat.html``

panpawn commented on chat-plugins/trivia.js:
```diff
@@ -530,7 +524,7 @@ class Trivia extends Rooms.RoomGame {
 		clearTimeout(this.phaseTimeout);
 		this.phaseTimeout = null;
-		this.broadcast('The trivia game was forcibly ended by ' + Chat.escapeHTML(user.name) + '.');
+		this.broadcast(`The trivia game was forcibly ended by ${Chat.escapeHTML(user.name)}.`);
```
This should also use ``Chat.html``

panpawn commented on chat-plugins/trivia.js:
```diff
@@ -549,15 +543,15 @@ class FirstModeTrivia extends Trivia {
 
-		let buffer = 'Correct: ' + Chat.escapeHTML(user.name) + '<br />' +
-			'Answer(s): ' + this.curAnswers.join(', ') + '<br />';
+		let buffer = `Correct: ${Chat.escapeHTML(user.name)}<br />` +
```
``Chat.html``

panpawn commented on chat-plugins/trivia.js:
```diff
@@ -1368,12 +1352,649 @@ module.exports = {
+		buffer += Object.keys(this.players)
+			.map(p => {
+				let player = this.players[p];
+				return `${Chat.escapeHTML(player.name)}${this.finals ? " (" + player.correctAnswers + ")" : ''}`;
```
``Chat.html``

panpawn commented on chat-plugins/trivia.js:
```diff
@@ -1368,12 +1352,649 @@ module.exports = {
+			buffer += Object.keys(this.players).sort()
+				.map(userID => {
+					let player = this.players[userID];
+					return `<button class="button" name="send" value="/wl vote ${player.userid}">${Chat.escapeHTML(player.name)}</button>`;
```
``Chat.html``

panpawn commented on chat-plugins/trivia.js:
```diff
@@ -1368,12 +1352,649 @@ module.exports = {
+		}
+		buffer += "</div>";
+		if (this.state === 'beginround') {
+			buffer += `<hr>The ${this.finals ? 'final' : 'next'} round will begin in 30 seconds. It will be <em style="font-weight: bold">${Chat.escapeHTML(this.curPlayer.name)}</em>'s turn to answer.`;
```
``Chat.html``

panpawn commented on chat-plugins/trivia.js:
```diff
@@ -1368,12 +1352,649 @@ module.exports = {
+			}
+		}
+		if (this.state === 'beforeround' || this.state === 'banking') {
+			this.sendToRoom(`|c:|${(Math.floor(Date.now() / 1000))}|~|The ${this.state === 'banking' ? 'next question' : (this.finals ? 'final round' : 'next round')} will begin in ${this.state === 'banking' ? '6' : '5'} seconds. It will be **${Chat.escapeHTML(this.curPlayer.name)}**'s turn to answer.`);
```
``Chat.html``

Zarel commented:
I'll let @Morfent decide when to approve

Morfent commented on chat-plugins/trivia.js:
```diff
@@ -54,7 +54,9 @@ if (!triviaData || typeof triviaData !== 'object') triviaData = {};
 if (typeof triviaData.leaderboard !== 'object') triviaData.leaderboard = {};
 if (!Array.isArray(triviaData.ladder)) triviaData.ladder = [];
 if (!Array.isArray(triviaData.questions)) triviaData.questions = [];
+if (!Array.isArray(triviaData.wlquestions)) triviaData.wlquestions = [];
```
Why do WL questions need to be differentiated from regular trivia questions?

Morfent commented on chat-plugins/trivia.js:
```diff
@@ -480,25 +481,18 @@ class Trivia extends Rooms.RoomGame {
-						} else {
-							ladder[rank] = [leader];
-						}
+				if (!i) {
```
Don't change the algorithm for this for now. It's something I'll get around to rewriting later, since it's really inefficient

Morfent commented on chat-plugins/trivia.js:
```diff
@@ -1258,43 +1251,44 @@ const commands = {
 		if (!ladder.length) return this.errorReply("No trivia games have been played yet.");
 
-		let buffer = "|raw|<div class=\"ladder\"><table>" +
+		let buffer = "|raw|<div class=\"ladder\" style=\"overflow-y: scroll; max-height: 300px;\"><table style=\"width: 100%\">" +
```
Why 300 pixels specifically?

CheeseMuffin replied:
it seemed to fit nicely was the main reason

Zarel replied:
`\"` triggers me; why are these not all `` ` `` strings?

CheeseMuffin replied:
Is it really worth replacing with template strings just to get rid of the \"?

sparkychildcharlie replied:
it's stylistically easier to read

panpawn replied:
That, and I mean, that is kind of what PS is going towards standards-wise

Zarel replied:
You don't need to replace all of them, but at least replace the ones you're touching.

Morfent commented on chat-plugins/trivia.js:
```diff
@@ -1258,43 +1251,44 @@ const commands = {
-
-		for (let i = 0; i < ladder.length; i++) {
+		let num = parseInt(target);
+		if (!num) num = 100;
```
Check if it's below zero too, since negative numbers are truthy

CheeseMuffin replied:
Oh yeah

Morfent commented:
I don't think there should be so much of a distinction between Trivia and Weakest Link. Treating them entirely differently like this makes the code handling both types of games more complex. Weakest Link should just be a game mode and `Weakest Link` should be subclassing `Trivia`, since its whole purpose is to provide an API for different types of trivia games like this while allowing them to handle rounds and scoring entirely differently. Weakest Link shouldn't need its own separate list of commands either, it should be able to work with Trivia's.

CheeseMuffin commented:
I guess I just don't see what exactly I would be able to reuse from the original API. The way handling questions are asked, handling guesses, won't be updating the leaderboard, the bank, makes it completely different from the way the current trivia games work. Azuuli originally said questions should be different, but that might end up changing, in which case the commands could be unified quite a bit. I did end up copying verifyAnswer, but besides that I'm not sure what can be reused from the current API.

Morfent commented on chat-plugins/trivia.js:
```diff
@@ -1368,12 +1359,649 @@ module.exports = {
+		super(room);
+		this.bank = 0;
+		this.amountToBank = 0;
+		if (room.gameNumber) {
```
This should be in `Trivia`'s constructor, even though it doesn't use this at the moment

Morfent commented:
First, Timer, and Number modes already asking questions, guesses, and scoring entirely differently. There's nothing wrong with overriding its methods to make it fit your needs better. The class doesn't make any reference to the leaderboard outside `Trivia#win`, which doesn't even belong in the class and will be removed once I get around to refactoring `triviaData`. I'm fairly positive `Trivia` is usable for this

The only problem that's making it not usable for this in its current state is that it makes some broadcasts meant specifically for the other modes and `.prize` has no actual purpose. I'll make a commit to fix this

Morfent commented:
https://github.com/Zarel/Pokemon-Showdown/commit/32c9bf8ae9df8396ccf2d1b87641c3a1916a745b

CheeseMuffin commented:
Updates:

- Uses Trivia API
- Most commands follow same format as trivia, only a few game-specific commands
- Fixed bug in onJoin
- Reverted to only ROs being able to see answers in qw

Does this look better @Morfent?

CheeseMuffin commented:
Not quite sure why I'm having so many problems with Travis, `npm test` is passing locally but then it keeps failing tests here...

panpawn commented:
Is your eslint updated?

CheeseMuffin commented:
ah it wasn't thanks

Morfent commented on chat-plugins/trivia.js:
```diff
@@ -1368,12 +1352,649 @@ module.exports = {
+		this.playerCap = 10;
+		this.order = [];
+		this.round = 0;
+		this.questions = Dex.shuffle(triviaData.wlquestions.slice());
```
Never mind, I assumed `Dex.shuffle` wouldn't whatever you pass to it

Morfent commented on chat-plugins/trivia.js:
```diff
@@ -836,16 +843,25 @@ const commands = {
 		// This prevents trivia games from modifying the trivia database.
 		questions = Object.assign([], questions);
 		// Randomizes the order of the questions.
 		questions = Dex.shuffle(questions);
```
`Object.assign([], questions)` was dumb of me to write and should be `questions.slice()`

Shuffle the questions at the end of the function so you're not shuffling both Trivia and WL's questions when it's not needed though.

Morfent commented on chat-plugins/trivia.js:
```diff
@@ -864,9 +882,9 @@ const commands = {
+			return this.errorReply(`There is already a game of ${room.game.title} in progress.`);
 		}
-
+		if (room.game.phase !== SIGNUP_PHASE && !room.game.canLateJoin) return this.errorReply("This game does not allow latejoins.");
```
`Trivia#addPlayer` should be checking this so commands don't get mixed up in how `Trivia` manages its state. The command can throw any string returned as an error reply and send the default signup message in the other case instead

Morfent commented on chat-plugins/trivia.js:
```diff
@@ -1358,12 +1675,17 @@ module.exports = {
 		trivia: commands,
 		ta: commands.answer,
-		// TODO: this is ugly as all hell, split it into sections
+		wl: {
```
`/wl` should just be an alias for `/trivia` as a whole instead of cherrypicking which commands to alias

CheeseMuffin commented:
Any other comments/concerns @morfent?

Morfent commented on chat-plugins/trivia.js:
```diff
@@ -1675,11 +1676,7 @@ module.exports = {
-			submit: commands.submit,
-			qs: commands.qs,
-		},
+		wl: commands,
```
`/wla` as an alias for `/wl answer` instead of having to use `/ta`would make more sense

panpawn replied:
... ``/wl`` is an alias for ``/weeklock`` :|

Morfent commented:
Apart from the one command, I don't think there's much else left to fix. I'll write some unit tests for this after merging
